File: 739-daily-temperatures/daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
//questions: are there negative temperatures?

        //things to keep in mind:
        //if there are no warmer day ahead res[idx]=0
        // track the preivous day temp using a stack
        int[] res = new int[temperatures.length];
        Stack<int[]> stack = new Stack<>(); //stores the value as [temp,idx]
    
        for(int i=0; i<temperatures.length; i++){
            int t=temperatures[i];
            while(!stack.isEmpty() && t>stack.peek()[0]){
                int[] pair = stack.pop();
                res[pair[1]]=i-pair[1];
        }
        stack.push(new int[]{t,i});
        }
        return res;
    }
}
    //stack [{73,0}]

File: 739-daily-temperatures/daily-temperatures.py
from typing import List
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n=len(temperatures)
        ans = [0] * n
        stack = []

        #run a loop through the temp 
        for i, temp in enumerate(temperatures):
            #run a loop to compare the current temp and TOS (high temp)
            while stack and temp > temperatures[stack[-1]]:
                # current temp is greater
                prev_idx = stack.pop()
                ans[prev_idx] = i - prev_idx
            #current temp is smaller, push to the stack
            stack.append(i)
        
        return ans

